airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py

```diff
@@ -0,0 +1,167 @@
+        """
+        Register custom normalization callback.
+        :param normalization_callback function to be used for value
+        normalization. Should return normalized value.
+        :return Same callbeck, this is usefull for using register function as decorator.
+        """
+        if TransformConfig.CustomSchemaNormalization not in self._config:
+            raise Exception("Please set TransformConfig.CustomSchemaNormalization config before registering custom normalizer")
+        self._custom_normalizer = normalization_callback
+        return normalization_callback
+
+    def __normalize(self, original_item: Any, subschema: Dict[str, Any]) -> Any:
+        """
+        Applies different transform function to object's field according to config.
+        :param original_item original value of field.
+        :param subschema part of the jsonschema containing field type/format data.
+        :return Final field value.
+        """
+        if TransformConfig.DefaultSchemaNormalization in self._config:
+            original_item = self.default_convert(original_item, subschema)
+
+        if self._custom_normalizer:
+            original_item = self._custom_normalizer(original_item, subschema)
+        return original_item
+
+    @staticmethod
+    def default_convert(original_item: Any, subschema: Dict[str, Any]) -> Any:
+        """
+        Default transform function that is used when TransformConfig.DefaultSchemaNormalization flag set.
+        :param original_item original value of field.
+        :param subschema part of the jsonschema containing field type/format data.
+        :return transformed field value.
+        """
+        target_type = subschema.get("type")
+        if original_item is None and "null" in target_type:
+            return None
+        if isinstance(target_type, list):
+            target_type = [t for t in target_type if t != "null"]
+            if len(target_type) != 1:
+                return original_item
+            target_type = target_type[0]
+        try:
+            if target_type == "string":
+                return str(original_item)
+            elif target_type == "number":
+                return float(original_item)
+            elif target_type == "integer":
+                return int(original_item)
+            elif target_type == "boolean":
+                return bool(original_item)
+        except ValueError:
+            return original_item
+        return original_item
+
+    def __normalize_and_validate(self, schema_key: str, original_validator: Callable):
+        """
+        Traverse through object fields using native jsonschema validator and apply normalization function.
+        :param schema_key related json schema key that currently being validated/normalized.
+        :original_validator: native jsonschema validator callback.
+        """
+
+        def normalizator(validator_instance, val, instance, schema):
```

Comment:
would be useful to explain the structure and purpose of this function. 

As I understand this requires these parameters since it is a jsonschema validator that we are passing into [validators.create](https://python-jsonschema.readthedocs.io/en/stable/creating/?highlight=validators.create#jsonschema.validators.create) but this isn't immediately clear in the code without hunting for the jsonschema docs.

Reply:
Yes, added a comment